GPU Encoding and Bump Estimation Performance

armansito opened the pull request:
- Performance data and discussion for CPU encoding, bump buffer estimation, GPU tag monoid prefix sums.
- Fixed the benchmark tool to warm up the CPU to increase the clock rate.
- Added to Future Work section.
- Cleaned up TODOs.
- Removed obsolete subsection placeholders. Might still write up the subsection on the flattening algorithm if time permits.

armansito committed:
[results] Benchmarks for input processing, encoding with bump estimate

armansito committed:
[bench] code to measure input processing

armansito committed:
input processing times on RTX 4090

armansito committed:
GPU encoding perf section

armansito committed:
Rebase & make paper

armansito committed:
Fixup mmark names; more content; future work

armansito committed:
Fix bug on paper.bib

armansito committed:
Change nehab_gpu_timings y-axis units to milliseconds

armansito committed:
fix tag_monoid_gpu_timings axis labels

armansito committed:
use log scale for the tag monoid plot

armansito committed:
fix LaTeX

raphlinus reviewed (approved):
I'm still skeptical about the encode timings on waves, but overall this looks good to me.

armansito committed:
Add a actual vs estimated memory table (which I forgot to do earlier) and expand discussion

armansito commented:
> I'm still skeptical about the encode timings on waves, but overall this looks good to me.

Yes, the timing for waves was wrong as I forgot about the fact that SVG scene fragments are cached. I updated Table 3 to reflect the correct timings. I also re-computed averages for the same table.

I also added Table 4 which contains actual and estimated bump buffer sizes. The text was referring to such a table where it discussed overestimation by Wang's but was linking to Table 3 which is about timing. I also expanded the discussion on memory estimation, elaborated on long dash scene encode timings and CPU-side dashing.

I'll leave this PR up for you to take another look before merging. The GPU results section feels complete now but please sign off on it.

armansito committed:
add missing units

armansito committed:
fix MotionMark citation

raphlinus commented on paper.tex:
```diff
@@ -522,30 +515,69 @@ \subsection{GPU: Execution Time} \label{subsection:gpu-results}
 \begin{itemize}
     \item The GPU implementation sequentializes some work that could be parallel, and is thus not as well load-balanced as it might be. An appealing future direction is to split the pipeline into separate stages, executed by the GPU as nodes in a work graph\cite{Patel2024}. This structure is appealing because load balancing is done by hardware.
 
+    \item The pre-allocation requirement for bump-allocated line soup buffer is a limitation due to today's graphics APIs. A more accurate and optimized buffer size estimation heuristic is considered future work. Another interesting future exploration is to employ newer GPU API features, such as mesh shaders and work graphs, for GPU-driven scheduling of the compute workload under a bounded memory ...
```
My main hesitation is that I don't see either mesh shaders or work graphs actually solving the GPC problem. I'd rewrite this:

\item Today's graphics APIs do not support dynamic reallocation from within a shader, thus force us to pre-allocate the bump-allocated line soup buffer. A more accurate and optimized buffer size estimation heuristic is considered future work. It may also be interesting to explore whether graphics APIs can be extended to facilitate GPU-driven scheduling of the compute workload under a bounded memory footprint.

I think it's also worth stating, probably earlier, that CUDA *can* reallocate, making it clear that this really is a graphics API specific problem.

armansito replied:
Done.

raphlinus reviewed (approved):
This all looks good. I have one suggestion about future work, but I think we need to go over that.

armansito committed:
rebase and address comments

armansito committed:
fix phrasing
